From 9eb9188f2ef27a8924fb8c4461908e7e15d0c2c1 Mon Sep 17 00:00:00 2001
Message-Id: <9eb9188f2ef27a8924fb8c4461908e7e15d0c2c1.1432852048.git.chang-joon.lee@intel.com>
In-Reply-To: <c5e9fa1af945ac4e7bd37c3035c0e756494f0273.1432852048.git.chang-joon.lee@intel.com>
References: <c5e9fa1af945ac4e7bd37c3035c0e756494f0273.1432852048.git.chang-joon.lee@intel.com>
From: John Harrison <John.C.Harrison@Intel.com>
Date: Tue, 5 May 2015 12:33:31 +0100
Subject: [PATCH 4/4] FOR_UPSTREAM [VPG]: drm/i915: Optimise out waits for
 completed requests

There is a function for synchronising objects across different rings.
This uses the object's 'last_read_req' request structure to determine
the last used ring. If different to the target ring, that request is
waited for in some manner (hardware semaphore, driver stall or
scheduler management). It was observed that this code was being called
quite frequently with the request having already been signalled as
completed. In this case, no further processing is required as there is
nothing to wait for.

This patch adds an early check for completion and bails out if found.

Change-Id: I280c5ad40e71a6f64905540bed32edcd132049fa
For: VIZ-5566
Signed-off-by: John Harrison <John.C.Harrison@Intel.com>
---
 drivers/gpu/drm/i915/i915_gem.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/gpu/drm/i915/i915_gem.c b/drivers/gpu/drm/i915/i915_gem.c
index cf9ee70..5d084d3 100644
--- a/drivers/gpu/drm/i915/i915_gem.c
+++ b/drivers/gpu/drm/i915/i915_gem.c
@@ -3282,6 +3282,9 @@ i915_gem_object_sync(struct drm_i915_gem_object *obj,
 		return 0;
 	}
 
+	if (i915_gem_request_completed(obj->last_read_req))
+		return 0;
+
 	if (to == NULL)
 		goto wait;
 
-- 
1.9.1

